File: src/main/java/ca/weblite/jdeploy/app/controllers/MainMenuViewController.java
package ca.weblite.jdeploy.app.controllers;

import ca.weblite.jdeploy.DIContext;
import ca.weblite.jdeploy.app.factories.ControllerFactory;
import ca.weblite.jdeploy.app.forms.MainMenuForm;
import ca.weblite.jdeploy.app.records.Project;
import ca.weblite.jdeploy.app.services.Edt;
import ca.weblite.jdeploy.app.services.ProjectService;
import ca.weblite.jdeploy.app.swing.ResponsiveImagePanel;
import ca.weblite.jdeploy.app.views.mainMenu.ProjectListCellRenderer;

import javax.swing.*;
import javax.swing.event.DocumentListener;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;

public class MainMenuViewController extends JFrameViewController {

    private final ProjectService projectService;
    private final ControllerFactory controllerFactory;

    public MainMenuViewController() {
        this.projectService = DIContext.get(ProjectService.class);
        this.controllerFactory = DIContext.get(ControllerFactory.class);
    }
    @Override
    protected JComponent initUI() {
        Edt edt = DIContext.get(Edt.class);

        MainMenuForm mainMenu = new MainMenuForm();

        mainMenu.getOpenButton().addActionListener(e -> {
            edt.invokeLater(new OpenProjectController(getFrame()));
        });

        mainMenu.getRecentProjects().setModel(buildRecentProjectsModel());
        mainMenu.getRecentProjects().setCellRenderer(new ProjectListCellRenderer());

        Action openRecentAction = openRecentAction(mainMenu);

        // openRecentAction should only be enabled when there is a project in the recent projects list selected
        mainMenu.getRecentProjects().getSelectionModel().addListSelectionListener(e -> {
            openRecentAction.setEnabled(mainMenu.getRecentProjects().getSelectedValue() != null);
        });
        mainMenu.getOpenRecentButton().setAction(openRecentAction);

        mainMenu.getRecentProjects().addMouseListener(new MouseAdapter() {
            @Override
            public void mouseClicked(MouseEvent e) {
                if (e.getClickCount() == 2) {
                    openRecentAction.actionPerformed(null);
                }
            }
        });
        mainMenu.getSearchField().getDocument().addDocumentListener(new DocumentListener() {
            @Override
            public void insertUpdate(javax.swing.event.DocumentEvent e) {
                filterProjects();
            }

            @Override
            public void removeUpdate(javax.swing.event.DocumentEvent e) {
                filterProjects();
            }

            @Override
            public void changedUpdate(javax.swing.event.DocumentEvent e) {
                filterProjects();
            }

            private void filterProjects() {
                String query = mainMenu.getSearchField().getText();
                DefaultListModel<Project> model = (DefaultListModel<Project>) mainMenu.getRecentProjects().getModel();
                model.clear();
                for (Project project: projectService.findRecent()) {
                    if (!query.isEmpty() && !project.getName().toLowerCase().contains(query.toLowerCase())) {
                        continue;
                    }
                    model.addElement(project);
                }
            }
        });

        mainMenu.getImportProject().addActionListener(e -> {
            new ImportProjectViewController(getFrame()).run();
        });

        mainMenu.getCreateProjectButton().addActionListener(e -> {
            var newProjectController = new NewProjectController(getFrame());
            newProjectController.show();
        });

        mainMenu.getHeroGraphicWrapper().add(
                new ResponsiveImagePanel(
                        "/ca/weblite/jdeploy/app/assets/jdeploy-home-hero.png"
                )
        , BorderLayout.CENTER);

        return mainMenu;
        //return new TestForm().getMhPanel();
    }

    @Override
    protected void onBeforeShow() {
        getFrame().setTitle("jDeploy");
    }

    private ListModel<Project> buildRecentProjectsModel() {
        DefaultListModel<Project> model = new DefaultListModel<>();
        for (Project project: projectService.findRecent()) {
            model.addElement(project);
        }

        return model;
    }

    private Action openRecentAction(MainMenuForm mainMenu) {
        return new AbstractAction() {

            {
                putValue(Action.NAME, "Open");
                putValue(Action.SHORT_DESCRIPTION, "Open the selected recent project");
                setEnabled(false);
            }
            @Override
            public void actionPerformed(ActionEvent e) {
                Project project = mainMenu.getRecentProjects().getSelectedValue();
                if (project != null) {
                    new OpenProjectController(getFrame(), project.getPath()).run();
                }
            }
        };
    }
}

File: src/main/java/ca/weblite/jdeploy/app/controllers/MainMenuViewController.kt
package ca.weblite.jdeploy.app.controllers

import ca.weblite.jdeploy.DIContext
import ca.weblite.jdeploy.app.factories.ControllerFactory
import ca.weblite.jdeploy.app.forms.AboutDialog
import ca.weblite.jdeploy.app.forms.MainMenuForm
import ca.weblite.jdeploy.app.records.Project
import ca.weblite.jdeploy.app.services.Edt
import ca.weblite.jdeploy.app.services.PreferencesService
import ca.weblite.jdeploy.app.services.ProjectService
import ca.weblite.jdeploy.app.swing.ResponsiveImagePanel
import ca.weblite.jdeploy.app.views.mainMenu.ProjectListCellRenderer
import ca.weblite.ktswing.coroutines.SwingDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import java.awt.BorderLayout
import java.awt.event.ActionEvent
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import java.util.*
import javax.swing.*
import javax.swing.event.DocumentEvent
import javax.swing.event.DocumentListener
import javax.swing.event.ListSelectionEvent

class MainMenuViewController : JFrameViewController() {
    private val projectService: ProjectService = DIContext.get(ProjectService::class.java)
    private val controllerFactory: ControllerFactory = DIContext.get(ControllerFactory::class.java)
    private val preferencesService: PreferencesService = DIContext.get(PreferencesService::class.java)

    companion object {
        const val MCP_TOOLS_ENABLED_KEY = "mcpToolsEnabled"
    }

    override fun initUI(): JComponent {
        val edt = DIContext.get(Edt::class.java)

        val mainMenu = MainMenuForm()

        mainMenu.getOpenButton().addActionListener { e: ActionEvent? ->
            edt.invokeLater(
                OpenProjectController(
                    frame
                )
            )
        }

        mainMenu.getRecentProjects().model = buildRecentProjectsModel()
        mainMenu.getRecentProjects().cellRenderer = ProjectListCellRenderer()

        val openRecentAction = openRecentAction(mainMenu)

        // openRecentAction should only be enabled when there is a project in the recent projects list selected
        mainMenu.getRecentProjects().selectionModel.addListSelectionListener { e: ListSelectionEvent? ->
            openRecentAction.isEnabled = mainMenu.getRecentProjects().selectedValue != null
        }
        mainMenu.getOpenRecentButton().action = openRecentAction

        mainMenu.getRecentProjects().addMouseListener(object : MouseAdapter() {
            override fun mouseClicked(e: MouseEvent) {
                if (e.clickCount == 2) {
                    openRecentAction.actionPerformed(null)
                }
            }
        })
        mainMenu.getSearchField().document.addDocumentListener(object : DocumentListener {
            override fun insertUpdate(e: DocumentEvent) {
                filterProjects()
            }

            override fun removeUpdate(e: DocumentEvent) {
                filterProjects()
            }

            override fun changedUpdate(e: DocumentEvent) {
                filterProjects()
            }

            fun filterProjects() {
                val query = mainMenu.getSearchField().text
                val model = mainMenu.getRecentProjects().model as DefaultListModel<Project>
                model.clear()
                for (project in projectService.findRecent()) {
                    if (!query.isEmpty() && !project.name.lowercase(Locale.getDefault())
                            .contains(query.lowercase(Locale.getDefault()))
                    ) {
                        continue
                    }
                    model.addElement(project)
                }
            }
        })

        mainMenu.getImportProject().addActionListener { e: ActionEvent? ->
            ImportProjectViewController(
                frame
            ).run()
        }

        mainMenu.getCreateProjectButton().addActionListener { e: ActionEvent? ->
            val newProjectController = NewProjectController(frame)
            // create coroutine scope
            val scope = CoroutineScope(SwingDispatcher)
            scope.launch {
                // show the new project controller
                newProjectController.show()
            }
        }

        mainMenu.getHeroGraphicWrapper().add(
            ResponsiveImagePanel(
                "/ca/weblite/jdeploy/app/assets/jdeploy-home-hero.png"
            ),
            BorderLayout.CENTER
        )

        // MCP tools checkbox - load from preferences (default: enabled)
        val rootPrefs = preferencesService.rootPreferences
        val mcpEnabled = rootPrefs.get(MCP_TOOLS_ENABLED_KEY, "true") == "true"
        mainMenu.getMcpToolsCheckBox().isSelected = mcpEnabled
        mainMenu.getMcpToolsCheckBox().addActionListener {
            val enabled = mainMenu.getMcpToolsCheckBox().isSelected
            rootPrefs.set(MCP_TOOLS_ENABLED_KEY, enabled.toString())
            try {
                rootPrefs.commit()
            } catch (_: Exception) {}
        }

        return mainMenu
        //return new TestForm().getMhPanel();
    }

    override fun onBeforeShow() {
        frame.title = "jDeploy"
        setupMenuBar()
    }
    
    private fun setupMenuBar() {
        val menuBar = JMenuBar()
        
        // Help menu
        val helpMenu = JMenu("Help")
        
        // About menu item
        val aboutMenuItem = JMenuItem("About jDeploy")
        aboutMenuItem.addActionListener { e: ActionEvent? ->
            val aboutDialog = AboutDialog(frame)
            aboutDialog.isVisible = true
        }
        
        helpMenu.add(aboutMenuItem)
        menuBar.add(helpMenu)
        
        frame.jMenuBar = menuBar
    }

    private fun buildRecentProjectsModel(): ListModel<Project> {
        val model = DefaultListModel<Project>()
        for (project in projectService.findRecent()) {
            model.addElement(project)
        }

        return model
    }

    private fun openRecentAction(mainMenu: MainMenuForm): Action {
        return object : AbstractAction() {
            init {
                putValue(NAME, "Open")
                putValue(SHORT_DESCRIPTION, "Open the selected recent project")
                isEnabled = false
            }

            override fun actionPerformed(e: ActionEvent?) {
                val project = mainMenu.getRecentProjects().selectedValue
                if (project != null) {
                    OpenProjectController(frame, project.path).run()
                }
            }
        }
    }
}
